lib/sdl.cpp

```diff
@@ -0,0 +1,804 @@
+         }
+         break;
+      case SDL_WINDOWEVENT_EXPOSED:
+         if (onExpose)
+         {
+            wnd_state = RenderState::ExposePending;
+         }
+         break;
+      default:
+         break;
+   }
+}
+
+void SdlWindow::motionEvent(SDL_MouseMotionEvent& em)
+{
+   EventInfo info =
+   {
+      em.x, em.y,
+      SDL_GetModState()
+   };
+   if (em.state & SDL_BUTTON_LMASK)
+   {
+      if (onMouseMove[SDL_BUTTON_LEFT])
+      {
+         onMouseMove[SDL_BUTTON_LEFT](&info);
+      }
+   }
+   else if (em.state & SDL_BUTTON_RMASK)
+   {
+      if (onMouseMove[SDL_BUTTON_RIGHT])
+      {
+         onMouseMove[SDL_BUTTON_RIGHT](&info);
+      }
+   }
+   else if (em.state & SDL_BUTTON_MMASK)
+   {
+      if (onMouseMove[SDL_BUTTON_MIDDLE])
+      {
+         onMouseMove[SDL_BUTTON_MIDDLE](&info);
+      }
+   }
+}
+
+void SdlWindow::mouseEventDown(SDL_MouseButtonEvent& eb)
+{
+   if (onMouseDown[eb.button])
+   {
+      EventInfo info =
+      {
+         eb.x, eb.y,
+         SDL_GetModState()
+      };
+      onMouseDown[eb.button](&info);
+   }
+}
+
+void SdlWindow::mouseEventUp(SDL_MouseButtonEvent& eb)
+{
+   if (onMouseUp[eb.button])
+   {
+      EventInfo info =
+      {
+         eb.x, eb.y,
+         SDL_GetModState()
+      };
+      onMouseUp[eb.button](&info);
+   }
+}
+
+void SdlWindow::keyEvent(SDL_Keysym& ks)
+{
+   bool handled = false;
+   if (ks.sym >= 128 || ks.sym < 32)
+   {
+      if (onKeyDown[ks.sym])
+      {
+         onKeyDown[ks.sym](ks.mod);
+         handled = true;
+      }
+   }
+   else if (ctrlDown)
+   {
+      if (onKeyDown[ks.sym])
+      {
+         onKeyDown[ks.sym](ks.mod);
+         handled = true;
+      }
+   }
+   if (ks.sym == SDLK_RCTRL || ks.sym == SDLK_LCTRL)
+   {
+      ctrlDown = true;
+   }
+   if (handled)
+   {
+      bool isAlt = ks.mod & (KMOD_ALT);
+      bool isCtrl = ks.mod & (KMOD_CTRL);
+      saved_keys += "[";
+      if (isCtrl) { saved_keys += "C-"; }
+      if (isAlt) { saved_keys += "Alt-"; }
+      if (ks.sym < 256 && std::isalpha(ks.sym))
+      {
+         //key with corresponding text output
+         char c = ks.sym;
+         if (!(ks.mod & KMOD_SHIFT)) { c = std::tolower(c); }
+         saved_keys += c;
+      }
+      else
+      {
+         saved_keys += SDL_GetKeyName(ks.sym);
+      }
+      saved_keys += "]";
+   }
+}
+
+void SdlWindow::keyEvent(char c)
+{
+   if (onKeyDown[c])
+   {
+      SDL_Keymod mods = SDL_GetModState();
+      bool isAlt = mods & (KMOD_ALT);
+      bool isCtrl = mods & (KMOD_CTRL);
+      onKeyDown[c](mods);
+      if (isAlt || isCtrl)
+      {
+         saved_keys += "[";
+         if (isCtrl) { saved_keys += "C-"; }
+         if (isAlt) { saved_keys += "Alt-"; }
+      }
+      saved_keys += c;
+      if (isAlt || isCtrl)
+      {
+         saved_keys += "]";
+      }
+   }
+}
+
+void SdlWindow::mainIter()
+{
+   SDL_Event e;
+   static bool useIdle = false;
+   if (SDL_PollEvent(&e))
```

Comment:
@v-dobrev, I think we need to go back to `while (SDL_PollEvent(&e))` here - there's noticeable response lag introduced when moving with mouse on a remote X11 connection to LC.

Reply:
I don't think this should block today's merge though.